### PR TITLE
Fixing: Deletion of Hibernated Clusters broken

### LINKED ISSUE #1089: Shoot stuck in deletion in a edge case
**What happened**:

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a Shoot and ensure that the creation passes successfully.
2. HIbernate the Shoot and ensure that all CP components are scaled down to 0.
3. Un-hibernate the shoot and trigger the deletion in the same time.
4. Ensure that the deletion stucks in `Initializing connection to Shoot` waiting kube-apiserver to be responsible but it is scaled down to 0.

**Anything else we need to know?**:

**Environment**:

- Gardener version: 0.23.0
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### LINKED ISSUE #1251: Shoot Spec: hibernation.enabled should be optional
**What would you like to be added**:
The field hibernation.enabled of the Shoot spec should be optional

**Why is this needed**:
I want to add a hibernation schedule to a cluster. Currently, I have to supply a value for `enabled`. As a result, the cluster is hibernated or woken up whenever the config is applied (depending on the value of enabled). This leads e.g. to an unexpected wake up of the cluster when enabled is set to false and the config is applied during hibernation time.

```
hibernation:
    schedules:
      - start: '0 15 * * 1,2,3,4,5'
        end: '0 8 * * 1,2,3,4,5'
        location: Europe/Berlin
```

### PR DESCRIPTION
**Pull request by danielfoehrKn**
**What this PR does / why we need it**:
This PR fixes the deletion of hibernated clusters by introducing the Status.isHibernated field in the Shoot.

Tested following scenarios:

1) Normal Hibernation
2) Deletion after being Hibernated
3) Waking up + Deletion at the same time when being already hibernated (did not enter normal reconcile yet)
4) Hibernate + Delete Shoot at the same time  (did not enter normal reconcile yet)
5) First hibernate, then set deletion timestamp while hibernating
6) Hibernating fails (hibernation status is not being set - only when successfully reconciled) -> then delete the shoot.   (Controle Plane already scaled down, spec.Hibernation.enabled == true, Status.isHibernated == nil)

**Which issue(s) this PR fixes**:
Fixes #1089
Fixes #1251

**Special notes for your reviewer**:

Added a timeout for the shoot deletion flow "Waiting until managed resources have been deleted".
During testing i ran twice into the issue that the previous step (delete the managed resource collection) did not set a deletion timestamp. 
Then the shoot deletion was stuck because it waited for the managed resources to be deleted. Needed a restart of the gardener controller manager to apply the deletion timestamp again.

However I could only reproduce the error in one case when I had a breakpoint set and then exceeded the context.  In the other case I could not find the reason for it. I ll keep an eye on it.

**Release note**:

```noteworthy user
The `.spec.hibernated.enabled` field in the `Shoot` resource is now optional.
```
```improvement user
The `Shoot` status does now contain a new `.status.hibernated` field which indicates whether the shoot is hibernated or not. The `.spec.hibernated.enabled` field just indicates the desired state of the user.
```

### CODE REVIEW

File: pkg/controllermanager/controller/shoot/shoot_care_control.go [hunk 1]
```diff
@@ -218,7 +218,7 @@ func garbageCollection(initShootClients func() error, botanist *botanistpkg.Bota
 		}
 	}()
 
-	if !botanist.Shoot.IsHibernated {
+	if !botanist.Shoot.HibernationEnabled {
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_care_control.go]: The health check should only be executed if `.status.hibernated=false`. Can you please do this?

File: pkg/controllermanager/controller/shoot/shoot_control_delete.go [hunk 2]
```diff
@@ -259,7 +258,7 @@ func (c *Controller) runDeleteShootFlow(o *operation.Operation) *gardencorev1alp
 		})
 		waitUntilManagedResourcesDeleted = g.Add(flow.Task{
 			Name:         "Waiting until managed resources have been deleted",
-			Fn:           botanist.WaitUntilManagedResourcesDeleted,
+			Fn:           flow.TaskFn(botanist.WaitUntilManagedResourcesDeleted).Timeout(10 * time.Minute),
```
Comment (rfranzke) [inline pkg/controllermanager/controller/shoot/shoot_control_delete.go:261]: nice, thanks for fixing that!

**Comment by rfranzke** [comment]
Also, please check the liniting/tests.

File: pkg/apis/garden/v1beta1/helper/helpers.go [hunk 3]
```diff
@@ -80,9 +80,9 @@ func GetShootCloudProvider(shoot *gardenv1beta1.Shoot) (gardenv1beta1.CloudProvi
 	return DetermineCloudProviderInShoot(shoot.Spec.Cloud)
 }
 
-// IsShootHibernated checks if the given shoot is hibernated.
+// IsShootHibernated checks if the given shoot's desired state is hibernated.
 func IsShootHibernated(shoot *gardenv1beta1.Shoot) bool {
```
Comment (timuthy) [inline pkg/apis/garden/v1beta1/helper/helpers.go]: Can you also rename this func like `HibernationDesired` or maybe you have something better in mind :)

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**